dulwich/graph.py

```diff
@@ -20,11 +20,73 @@
 
 """Implementation of merge-base following the approach of git."""
 
-from collections import deque
-from typing import Deque
+from collections.abc import MutableMapping
+from collections import OrderedDict
 
+from heapq import heappush, heappop
 
-def _find_lcas(lookup_parents, c1, c2s):
+# Because opening an sha file from the object_store, reading it,
+# then decoding its will be a slow operation, we want to create
+# a cache of Commit objects so that repeat parent and commit time
+# lookup (which are inherent when looking for a common ancestor)
+# will be much faster, so we need to implement a simple Commit cache
+# using an OrderedDict
+
+
+class CommitCache(MutableMapping):
```

Comment:
Won't this grow with the size of the graph? In other places we use LRUCache for this.

Reply:
It has a maxsize parameter which can be quite small as once the most recent commit works itself back to be nearer in time to the commit time of the older commit, the size of the Working List is quite small.

Reply:
Looked at the LRUCache implementation.

It could be used but is way overkill when a simple OrderedDict is all you need for a finite sized commit cache.  The version in this PR will perform much faster because it internally is using the built-in C implementation of OrderedDict and so does not need separate python based linked lists and their overhead.

But it is your call.  If you want this routine to use your LRUCache instead of being self-contained, faster/simpler, I will make that change.

Reply:
It's an existing and tested piece of infrastructure though. Why is an OrderedDict sufficient? What if the breadth is such that parents scroll out of view?

Either way, if you're adding a new class, please add unit tests specifically for that class.

Reply:
"What if the breadth is such that parents scroll out of view?" ... all fixed size LRU implementations has the same scope problem.  If you want LRUCache and slower that is fine by me.

Reply:
(the criteria for what scrolls out of view is different for this implementation vs the LRUCache though)

Yeah, I'd prefer that - I'd rather be conservative here and use whatever we already have. If it turns out there's a performance win to be had, let's measure implement that separately - and use it everywhere in Dulwich, rather than just here.

Reply:
The "criteria" is to add any new data to the end of the ordered (fifo) dict on writes, and on reads you move the most recently accessed to the just added end.  When cap is reached drop the first (oldest).  In other words an LRU algo with all linked lists and lookups handled internally by python's Ordered dict.  A pretty standard way to implement a LRU cache is with a  Ordered dict but will do as you ask since you will be the one to have to maintain it.

Reply:
fixed in my pr